### PR TITLE
Support for dropdown of predefined options in data sources setup

### PR DESCRIPTION
**Pull request by gabrieldutra**
## What type of PR is this? (check all applicable)

- [x] Feature

## Description
This adds support to use `Select` in Data sources/Destinations setup.

- [x] Rendering in frontend
- [x] Check if ath is missing backend-wise

## Related Tickets & Documents
--

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)
ds-predefined-options

### CODE REVIEW

File: redash/query_runner/impala_ds.py [hunk 1]
```diff
@@ -47,8 +47,12 @@ def configuration_schema(cls):
                     "type": "number"
                 },
                 "protocol": {
-                    "type": "string",
-                    "title": "Please specify beeswax or hiveserver2"
+                    "type": "enum",
+                    "options": [
+                        {"value": "beeswax"},
+                        {"value": "hiveserver2"}
+                    ],
```
Comment (arikfr) [inline redash/query_runner/impala_ds.py]: This won't pass schema validation. A valid enum is:

```json
{
    "protocol": {
        "type": "string",
        "enum": ["beeswax", "hiveserver2"]
    }
}
```

We might want to add our own extension to provide friendly names for options:

```
{
    "protocol": {
        "type": "string",
        "extendedEnum": [{"value", "beeswax", "name": "Beeswax"}, {"value": "hiveserver2", "name": "Hive Server 2"}]
    }
}
```

And we will transform `extendedEnum` into `enum` before passing it into `jsonschema.validate` call.

Comment (gabrieldutra) [reply redash/query_runner/impala_ds.py]: >We might want to add our own extension to provide friendly names for options

That's exactly why I went with the list of objects on that, but yeah, it does make sense to permit a simplified version.

> And we will transform extendedEnum into enum before passing it into jsonschema.validate call.

That's the only thing missing in terms of backend code?

**Comment by gabrieldutra** [comment]
> Will it pick some value by default? I think think it should pick either the default value as default and if no default is specified, then it should pick the first value.

For now only if you specify a default value, for other cases I went with the placeholder. First value is an option too, I'll update that.

**Comment by arikfr** [comment]
> For now only if you specify a default value, for other cases I went with the placeholder. First value is an option too, I'll update that.

Because it will fail validation if you don't specify a value from the list it's best if we just pick one for you.

File: redash/query_runner/impala_ds.py (hunk 1 above)
Comment (arikfr) [reply redash/query_runner/impala_ds.py]: > That's the only thing missing in terms of backend code?

I think so.

File: redash/utils/configuration.py [hunk 2]
```diff
@@ -24,6 +24,11 @@ def __init__(self, config, schema=None):
         self.set_schema(schema)
 
     def set_schema(self, schema):
+        if isinstance(schema, dict):
+            for prop in schema.get('properties', {}).values():
+                if 'extendedEnum' in prop:
+                    prop['enum'] = map(lambda v : v['value'], prop['extendedEnum'])
+                    del prop['extendedEnum']
```
Comment (gabrieldutra) [inline redash/utils/configuration.py]: @arikfr (or @rauchy)

This seemed the best place to put the transformation extendedEnum -> enum logic (so `ConfigurationContainer` stays in sync with what it's expected from jsonschema).

One note is that this modifies the schema in the upper scope (for the DS/Destination handler it didn't seem that big of a deal), so lmk if there is a dict "deep copy" that I should be using in such circumstances.

File: redash/utils/configuration.py [hunk 3]
```diff
@@ -24,6 +24,11 @@ def __init__(self, config, schema=None):
         self.set_schema(schema)
 
     def set_schema(self, schema):
+        if isinstance(schema, dict):
```
Comment (arikfr) [inline redash/utils/configuration.py]: This is not needed: `schema` is always a dict.

File: redash/utils/configuration.py (hunk 2 above)
Comment (arikfr) [reply redash/utils/configuration.py]: > One note is that this modifies the schema in the upper scope

:thinking: I wonder if we might end up in a situation where it returns the modified version in an API call. Several options:

1. Deep copy.
2. Keep `extendedEnum` and change the UI to use it first and ignore `enum` if `extendedEnum` exists.

File: redash/utils/configuration.py (hunk 3 above)
Comment (gabrieldutra) [reply redash/utils/configuration.py]: I was getting a "`NoneType` has no attribute..." And that's why I put it 😬

File: redash/utils/configuration.py (hunk 2 above)
Comment (gabrieldutra) [reply redash/utils/configuration.py]: There is already a preference to `extendedEnum` in UI, just need to keep it there then 😉

Comment (gabrieldutra) [reply redash/utils/configuration.py]: Deep copy seemed more future-proof, so I went with it in the end :+1:

**Comment by arikfr** [comment]
:+1:

**Pull request merged**